modules/plotting/src/interaction/boxselectioninteractionhandler.cpp

```diff
@@ -0,0 +1,205 @@
+    }
+    // For efficiency:
+    // 1. Determine selection along x-axis
+    // 2. Determine selection along y-axis using the subset from 1
+
+    auto xbuf = xAxis->getRepresentation<BufferRAM>();
+    auto selectedIndicesX = xbuf->dispatch<std::vector<size_t>, dispatching::filter::Scalars>(
+        [min = start[0], max = end[0]](auto brprecision) {
+            using ValueType = util::PrecisionValueType<decltype(brprecision)>;
+            std::vector<size_t> selectedIndices;
+            selectedIndices.reserve(brprecision->getSize());
+            for (auto&& [ind, elem] : util::enumerate(brprecision->getDataContainer())) {
+                if (static_cast<double>(elem) < min || static_cast<double>(elem) > max) {
+                    continue;
+                } else {
+                    selectedIndices.emplace_back(ind);
+                }
+            }
+            return selectedIndices;
+        });
+    // Use indices filted by x-axis as input
+    auto ybuf = yAxis->getRepresentation<BufferRAM>();
+    auto selectedIndices = ybuf->dispatch<std::unordered_set<size_t>, dispatching::filter::Scalars>(
+        [selectedIndicesX, min = start[1], max = end[1]](auto brprecision) {
+            using ValueType = util::PrecisionValueType<decltype(brprecision)>;
+            auto data = brprecision->getDataContainer();
+            std::unordered_set<size_t> selectedIndices;
+            selectedIndices.reserve(selectedIndicesX.size());
+
+            std::copy_if(selectedIndicesX.begin(), selectedIndicesX.end(),
+                         std::inserter(selectedIndices, selectedIndices.begin()),
+                         [data, min, max](auto idx) {
+                             return !(static_cast<double>(data[idx]) < min ||
+                                      static_cast<double>(data[idx]) > max);
+                         });
+
+            return selectedIndices;
+        });
+    return selectedIndices;
+}
+
+std::unordered_set<size_t> BoxSelectionInteractionHandler::boxFilter(const dvec2& start,
+                                                                     const dvec2& end,
+                                                                     const BufferBase* xAxis,
+                                                                     const BufferBase* yAxis) {
+    if (xAxis == nullptr || yAxis == nullptr) {
+        return std::unordered_set<size_t>();
+    }
+    auto xbuf = xAxis->getRepresentation<BufferRAM>();
+    std::unordered_set<size_t> filteredIndices =
+        xbuf->dispatch<std::unordered_set<size_t>, dispatching::filter::Scalars>(
+            [start, end, ybuf = yAxis->getRepresentation<BufferRAM>()](auto brprecision) {
+                using ValueType = util::PrecisionValueType<decltype(brprecision)>;
+                return ybuf->dispatch<std::unordered_set<size_t>, dispatching::filter::Scalars>(
+                    [start, end, xData = brprecision->getDataContainer()](auto brprecision) {
```

Comment:
think you are making a copy of the container here.

Reply:
Should return `const &`?

Reply:
yes, but `xData = ...`  will make a copy, need to use `&xData = ...` to make a ref...